Fix useFocusFirstElement in non-contenteditable Blocks

alexstine opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->

Fixes useFocusFirstElement to take in to account non-contenteditable Blocks. In this case, it performs a secondary search to find a more suitable focus point rather than inserting the caret at the edge which doesn't work well for Blocks without contenteditable.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

Tested in Firefox with NVDA screen reader on Windows 10.

1. Open the Block Editor.
2. Insert a Columns Block.
3. Focus is moved to the first focussable element which is the Block Variations panel.

## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

Enhancement

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->
- [x] I've updated related schemas if appropriate. <!-- Reference: https://github.com/WordPress/gutenberg/tree/trunk/schemas -->

alexstine commented on packages/block-editor/src/components/block-tools/block-selection-button.js:
```diff
@@ -190,7 +190,7 @@ function BlockSelectionButton( { clientId, rootClientId, blockElement } ) {
 				event.preventDefault();
 				selectBlock( focusedBlockUid );
-			} else if ( isTab && selectedBlockClientId ) {
+			}/* else if ( isTab && selectedBlockClientId ) {
```
What is this line for? It seems to not be working properly. Even checked out on trunk this line is an issue. If I have one Paragraph and one Columns, I will enter Gutenberg Navigation Mode. If I press Tab, I will land on Paragraph, then Columns, then I get the first button within Columns. After commenting this line, I get placed in the Title field and I can go through my list again which seems like a much better workflow than random focus of Columns block content and possibly other Blocks.

alexstine replied:
I am trying to fix this in the below PR.

https://github.com/WordPress/gutenberg/pull/37965

alexstine commented on packages/block-library/src/columns/edit.js:
```diff
@@ -240,29 +243,82 @@ function Placeholder( { clientId, name, setAttributes } ) {
+				>
+					{ accessibleTitle }
+				</Button>
+				{ isAccessibleModalOpen && (
```
I really don't like this. I am duplicating too much code I think. Would be cool at some point to figure out how to pull this content in to the Modal via a ref or something. However, that may not be much more efficient.

Mamaduka replied:
Hi, @alexstine

We could store `__experimentalBlockVariationPicker` into a variable then use it in both places. What do you think?

Example:

```js
const variationPicker = (
  <__experimentalBlockVariationPicker
	  icon={ get( blockType, [ 'icon', 'src' ] ) }
	  label={ get( blockType, [ 'title' ] ) }
	  variations={ variations }
	  onSelect={ ( nextVariation = defaultVariation ) => {
		  if ( nextVariation.attributes ) {
			  setAttributes( nextVariation.attributes );
		  }
		  if ( nextVariation.innerBlocks ) {
			  replaceInnerBlocks(
				  clientId,
				  createBlocksFromInnerBlocksTemplate(
					  nextVariation.innerBlocks
				  ),
				  true
			  );
		  }
	  } }
	  allowSkip
  />
);
```

alexstine reviewed (commented):
Initial thoughts. It is much more accessible but this will need a ton of work I think.

alexstine committed:
Finally fix selector in test I think.

MarcoZehe commented:
Sorry about that.

tellthemachines reviewed (commented):
Testing with VoiceOver on Safari, the dialog provides a much better experience than the default block placeholder. It does require markup duplication though, so it would be nicer if we could just make the existing placeholder markup more accessible. 

Comparing the dialog and the placeholder markup, the main difference seems to be in the use of `aria-labelledby` and `aria-describedby` in the dialog, that causes the label and description to be read. The placeholder markup is totally devoid of semantics, and VoiceOver ignores it.

The other aspect is, of course, that the dialog keeps focus inside it so users can't tab away from the placeholder before selecting an option. How important is this? Because if we don't have to constrain focus to the placeholder, we can experiment with other types of markup, for instance, putting the whole placeholder inside a fieldset, with the label and description as its legend.

alexstine commented:
@tellthemachines It was necessary to use the dialog because of the tabbing issues. Remember, for Windows users, Tab and Shift+Tab are primary navigation keys unlike the majority of cases on Mac with Voiceover. If there is some other approach that would allow us to get the same results of a dialog without using a dialog, that would be great. There are a ton of libraries at play in block-list though and it will be tough to figure out how to solve this without breaking it further. Essentially what needs to happen is this.

https://github.com/WordPress/gutenberg/pull/37965

The only difference being we need to only focus the Sidebar when there is no more focusable element in the block content. This could get tricky though playing around with focus which is why I actually do like the dialog better but I understand how intrusive it is.

My thoughts:
1. Insert a Columns Block.
2. Columns Block is presented as an application instead of document.
3. Add aria-labelledby for Columns title and aria-describedby for any needed descriptions.
4. Focus is placed on first Block Variation.
5. At this point, Shift+Tab should focus the Block Toolbar since there is no previous focusable element within the current content.
6. Tab should continue through the Block Variations until it reaches the last focusable element in the content, then it should land on Block Sidebar.

Any of this sounding reasonable? It just sounds insane to have to mess with focus like this.

tellthemachines commented:
> It was necessary to use the dialog because of the tabbing issues. Remember, for Windows users, Tab and Shift+Tab are primary navigation keys unlike the majority of cases on Mac with Voiceover.

I'm not sure I've got this 100%, so to clarify: do we want to move focus into the block placeholder with the Tab key, when a new block is inserted? Currently, when inserting a Columns block, focus moves to the whole block, and then we have to press Down Arrow in order to access the buttons inside. But once inside, all the controls can be navigated with Tab.

Would focusing the first focusable element in the block after insertion solve the problem? Plus adding semantic markup so the block heading and description are read out, of course.

Whatever we decide on, the solution should work for all blocks that have placeholders: Navigation, Cover, Image, Gallery, Table, etc. so ideally it should be a global change in the placeholder, as opposed to individual changes for each block.

alexstine commented:
@tellthemachines Oh no, I forgot to test this.

So, you are correct in one case.
1. Insert a Columns Block.
2. Down Arrow will focus the Block Variations.
3. In theory, we should automatically focus the placeholder.

Here is the problem I am seeing.
1. Select Options > Preferences.
2. Select Blocks.
3. Enable Contain text cursor inside block.
4. Close the dialog and Options menu.
5. Insert a Columns Block.
6. Notice how pressing Down Arrow does not focus the Block Variations panel and how focus jumps directly to the Block Settings Sidebar.

Sorry for not being more clear about this, but it works differently based on that setting. We've got to have this working the same way across the board.

Thanks.

alexstine commented:
It almost seems like Writing Flow is allowing some type of focus switching and whenever Writing Flow is disabled for switching from Block to Block, this is why it breaks.

tellthemachines commented:
Oooh, I should have thought to test it with the "Contain text cursor inside block" setting! It makes sense that Arrow key navigation wouldn't work with that setting enabled.

I'm thinking ultimately the only way we can make things work properly is making the whole editor content navigable with Tab. But perhaps to start with, we can make it so the block placeholder buttons can be accessed with Tab instead of Down Arrow.

alexstine committed:
If not contenteditable Block, find next focussable element.

alexstine commented:
@tellthemachines I just pushed a fix which should help with this. Thoughts? I tested with Columns and Navigation Blocks and Tab works fine now. I also gave the Paragraph Block a test just to ensure this didn't break existing functionality. Finally, refreshed the PR with latest commits.

This was a much easier solution, wish I had thought of this earlier.

Thanks.

tellthemachines reviewed (commented):
This works well! It solves the problem for blocks with placeholders when inserted inline. To fix it for blocks inserted with the header toolbar inserter will be a different problem, because those don't receive focus at all when inserted.

There are some e2e test failures on this branch, let's see if we can fix them and get this in.

alexstine committed:
Try to make tests stable.

alexstine committed:
Try to fix tests.

alexstine commented:
@tellthemachines Can you spot why the checks are failing? I have run them all locally and they all pass. I don't understand why these are failing this time...

alexstine committed:
Update snapshot.

alexstine committed:
Fix focus getting stuck on child Block wrapper. Adjust to find only specific whitelisted element.

alexstine commented on packages/block-editor/src/components/block-list/use-block-props/use-focus-first-element.js:
```diff
@@ -95,6 +95,24 @@ export function useFocusFirstElement( clientId ) {
+		if ( ! ref.current.getAttribute( 'contenteditable' ) ) {
+			const whitelistTags = [ 'a', 'button' ];
+			let focusElements = focus.tabbable.findNext( ref.current );
+			// In the event a Block has children, focus will find the next element with tabindex which would happen to be the wrapping element of a child Block. This ensures we only focus on whitelisted tag such as link/button.
```
What happened here was this.

1. Open the Block Editor.
2. Insert an Image Block.
3. Convert to Group Block from the Image Block Toolbar.
4. When focus was placed on the Group Block, the code found the next focussable element which happens to be the Image Block wrapper.
5. I adjusted the code so it must look for a whitelisted tag. This way focus will skip the Image Block wrapper and land on the first button/link contained within the Image Block.

I tried to sort this out with .filter() but I could just never get the node sorting to work properly.

tellthemachines replied:
When wrapping an image in a group, we might want focus to stay on the group, in order to customise it from the sidebar or something. In that case it's unexpected to have focus return to the image straightaway.

alexstine reviewed (commented):
Found another bug. Getting a little closer I hope.

alexstine committed:
Revert tests and start fresh.

alexstine committed:
More work on tests.

tellthemachines commented:
> @tellthemachines Can you spot why the checks are failing? I have run them all locally and they all pass. I don't understand why these are failing this time...

At least some of those tests are relying on the need to press Arrow Down to enter the block placeholder. [This one](https://github.com/WordPress/gutenberg/blob/trunk/packages/e2e-tests/specs/editor/various/writing-flow.test.js#L604-L605) for instance. I think here it's best to change the keyboard instructions in the test so the outcome is still whatever was originally intended.

alexstine committed:
Add td to whitelisted elements. Check to see if first element has focus potential. Ensure the element to focus is within the current Block. Try to make tests stable.

alexstine commented on packages/block-editor/src/components/block-list/use-block-props/use-focus-first-element.js:
```diff
@@ -95,6 +95,33 @@ export function useFocusFirstElement( clientId ) {
+			let firstFocusElements = true;
+			// In the event a Block has children, focus will find the next element with tabindex which would happen to be the wrapping element of a child Block. This ensures we only focus on whitelisted tag such as link/button.
+			do {
+				// The first element may be okay to focus. Go ahead and check.
```
In the event of the Table Block and maybe others, the first td could be a good point to focus. I changed the code slightly to always check the first element.

tellthemachines replied:
Table starts as a placeholder though, and its first focusable element is the "column count" input. I'm not sure a whitelist is the way to go here: apart from the risk a legitimate element may be left out, if the interface changes it might happen that the next focusable element after the block being inserted is a button or an anchor, and focus might end up in the wrong place.

alexstine commented on packages/block-editor/src/components/block-list/use-block-props/use-focus-first-element.js:
```diff
@@ -95,6 +95,33 @@ export function useFocusFirstElement( clientId ) {
+				} else {
+					focusElements = focus.tabbable.findNext( focusElements );
+				}
+				// Need to make sure we're still in the current Block, if not, we could run in to trouble if focus is placed in a Block further down the page.
```
I actually replicated this by mistake so I added a check to ensure the element was contained in the wrapping element of the Block. Otherwise, you could get dropped in a totally separate Block which would not be great.

tellthemachines reviewed (commented):
This has turned out a more complex piece of work than I initially thought. Some of those e2e failures are pointing to actual problems: there are a few blocks that should be directly editable - quote, code, HTML for instance - that are no longer focusing the first editable field when inserted. 

With these changes, it's also harder to delete a block immediately after insertion. If the whole block is selected, pressing Backspace will delete it. But with a placeholder element selected, an extra press of Arrow Up is needed to select the whole block, which won't work if the caret is contained to the block.

Perhaps we can explore other ways of doing this. Could we make "Contain text cursor inside block" work only when a contenteditable is focused? Alternatively we could try using Tab to enter the block, but that would require some rethinking of how Tab currently works in the editor.

alexstine commented:
> Could we make "Contain text cursor inside block" work only when a contenteditable is focused?

That won't fix our problem. It would only leave us with the same problem. Users would be able to navigate with Arrow Keys, but they may by mistake enter another Block. Then we've got to worry about communicating different interaction models. With that being said, would it be possible to tell this feature to be less picky? E.g. I have found that all Blocks have parent elements. Some of them are div, p, figure, etc. What if we allowed interaction inside the Block but still interaction could not happen outside of a Blocks wrapping element? That would be a really great solution.

>  Alternatively we could try using Tab to enter the block, but that would require some rethinking of how Tab currently works in the editor.

I think we should do this as well. I think Tab is fine to move to the Sidebar, but it should only work at the end of the Block content. E.g. If no more content to focus in Block, move to Sidebar. Shift+Tab from Sidebar will put you on last focus element within Block. However, I think we should add a dedicated shortcut for the Sidebar. E.g. there is already a shortcut to focus the Block Toolbar, why not a global shortcut for the Sidebar?

I think the focus on Block enter is an important thing to have but I think it is going to take a lot of work to get there. I will revert the code back to a previous state and let's see if we can get that portion in.

1. It checks for no contenteditable.
2. It checks to ensure the element is contained in the Block.
3. It executes focus if an element was found.

I think the biggest issue is not all Blocks can receive focus so that is something we can work on in another PR. E.g. the Group Block appears empty to me, we need to maybe add a focussable placeholder to every Block with instructions on how to interact with it or something that way users don't think it is empty and useless.

Does this make any sense? Any thoughts after my comment?

alexstine committed:
Revert some code.

alexstine committed:
Back to test fixing.

MarcoZehe commented:
> Alternatively we could try using Tab to enter the block, but that would require some rethinking of how Tab currently works in the editor.

There are several, at times conflicting, issues/suggestions/discussions about how tab currently works. Some want it to always be modal to the block, others want it to go from the block straight to the side bar, and ideally shift+tab would go directly back the same way, etc.

That's why in #38311 is there, to try and sort things out and see if concepts like Ctrl/Cmd+F6 (in the browser) to focus the side bar instead would help with untangling the mess that is the tab key currently.

alexstine committed:
Try again to prevent focus in child Blocks. Tests should pass this time, maybe.

alexstine committed:
Update another snapshot.

alexstine commented:
I think there may be something wrong with this test.

```
packages/e2e-tests/specs/editor/blocks/navigation.test.js#L798 
TimeoutError: waiting for function failed: timeout 30000ms exceeded
    at new WaitTask (/home/runner/work/gutenberg/gutenberg/node_modules/puppeteer-core/src/common/DOMWorld.ts:798:28)
    at DOMWorld.waitForFunction (/home/runner/work/gutenberg/gutenberg/node_modules/puppeteer-core/src/common/DOMWorld.ts:720:22)
    at Frame.waitForFunction (/home/runner/work/gutenberg/gutenberg/node_modules/puppeteer-core/src/common/FrameManager.ts:1352:28)
    at Page.waitForFunction (/home/runner/work/gutenberg/gutenberg/node_modules/puppeteer-core/src/common/Page.ts:3346:29)
    at waitForInserterCloseAndContentFocus (/home/runner/work/gutenberg/gutenberg/packages/e2e-test-utils/build/@wordpress/e2e-test-utils/src/inserter.js:92:8)
    at insertBlock (/home/runner/work/gutenberg/gutenberg/packages/e2e-test-utils/build/@wordpress/e2e-test-utils/src/inserter.js:171:8)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at Object.<anonymous> (/home/runner/work/gutenberg/gutenberg/packages/e2e-tests/specs/editor/blocks/navigation.test.js:249:4)
 Check failure on line 798 in packages/e2e-tests/specs/editor/blocks/navigation.test.js
```

Whenever I try testing this against trunk, it also fails.

1. Checkout trunk and npm run build.
2. Insert a Block from the global Inserter.
3. Type Navigation.
4. Press Enter on the Navigation Block to insert it.
5. Open the console and run the following.

```
document.activeElement.closest( '.block-editor-block-list__layout' );
```

6. You will get null returned.

It looks like this is not working unless I have some weird cache issue on my side.

alexstine committed:
More work on E2E tests.

alexstine commented on packages/block-editor/src/components/block-list/use-block-props/use-focus-first-element.js:
```diff
@@ -96,6 +96,27 @@ export function useFocusFirstElement( clientId ) {
+			}
+		}
+
 		setContentEditableWrapper( ref.current, false );
```
@ellatrix I am not sure what this line is for right off hand, is it okay to have my code run before this or do I need to switch the order?

alexstine reviewed (commented):
Question about order after merging latest changes from trunk.

talldan commented on packages/block-editor/src/components/block-list/use-block-props/use-focus-first-element.js:
```diff
@@ -96,6 +96,27 @@ export function useFocusFirstElement( clientId ) {
+				return;
+			}
+			// Need to try our best to detect child Blocks and not focus them straight away.
+			const disallowedTags = [ 'div', 'figure' ];
+			const checkDisallowedTags = disallowedTags.some(
+				( tag ) => focusElement.tagName.toLowerCase() === tag
+			);
```
If the aim is to detect child blocks, you could possibly check for the `data-block` attribute here, which all block wrappers should have.

It might get tricky if `focusElement` is an element within a child block, though.

alexstine replied:
@talldan That is the problem. I almost need to check a store somewhere to see if Block has children. Was trying to hack around it, but not very reliable.

talldan commented on packages/block-editor/src/components/block-list/use-block-props/use-focus-first-element.js:
```diff
@@ -96,6 +96,27 @@ export function useFocusFirstElement( clientId ) {
+			const checkDisallowedTags = disallowedTags.some(
+				( tag ) => focusElement.tagName.toLowerCase() === tag
+			);
+			if ( checkDisallowedTags ) {
+				focusElement = false;
+			}
+			if ( focusElement ) {
+				focusElement.focus();
+				return;
+			}
```
It'd be nice to keep it so that `focusElement` is always a consistent type. Could this be reduced to one `if` statement? e.g.

```js
if ( ! checkDisallowedTags && focusElement ) {
    focusElement.focus()
    return;
}
```

alexstine replied:
Done.

talldan reviewed (commented):
It's unfortunate that this kind of change sets off quite a few test failures. I'll have a look through to see if I can help with any of them.

I left a couple of code comments.

In testing, I noticed an issue with inserting the Group block that doesn't seem to happen in `trunk`. There's a console error when inserting it:
```js
Uncaught RangeError: Invalid array length
    at Module.findPrevious (tabbable.js:169:2)
    at preventScrollOnTab (use-tab-nav.js:185:19)
```

For most other blocks I tested is seems to work well.

I'm not sure which other contributors might also be familiar with this code and able to offer more insight. Maybe @ellatrix or @youknowriad?

talldan commented on packages/e2e-tests/specs/editor/various/__snapshots__/multi-block-selection.test.js.snap:
```diff
@@ -116,7 +116,7 @@ exports[`Multi-block selection should not multi select single block 1`] = `
 exports[`Multi-block selection should only trigger multi-selection when at the end 1`] = `
 "<!-- wp:paragraph -->
-<p>1.</p>
+<p>4.</p>
```
It seems like this value should go back to `1`. When I tried the test steps manually that was the result.

alexstine replied:
Fixed.

talldan commented on packages/e2e-tests/specs/editor/various/list-view.test.js:
```diff
@@ -82,11 +82,12 @@ describe( 'List view', () => {
 		await page.keyboard.press( 'Enter' );
 
 		const canvasImageBlock = await page.waitForSelector(
-			'figure[aria-label="Block: Image"]'
+			'button.block-editor-media-placeholder__upload-button'
 		);
 		expect( canvasImageBlock ).toHaveFocus();
```
It'd be good to rename the variable to `uploadButton`. There's also a preference on the project for using labels in selectors, though here that would involve an xpath selector. Something like this:
```js
const uploadButton = await page.waitForXPath(
	'//button[.="Upload"]'
);
```

alexstine replied:
I had no idea how to do that so I matched against the class. I thought I had to grab innerText or something. I've never seen xpath syntax before.

alexstine replied:
I further refined your xpath syntax as the one you suggested failed. Does it look good?

talldan replied:
Looks good 👍 

Xpaths have a tricky syntax, but sometimes they're the best option.

talldan reviewed (commented):
It looks like the navigation.test.js failures are related to this test utility function:
https://github.com/WordPress/gutenberg/blob/04ba33ee9585e7864e241313146526055e68ab2a/packages/e2e-test-utils/src/inserter.js#L163-L172

`focusSelectedBlock` no longer seems to work, the test expects that focusing the selected block will close the inserter sidebar, but I don't think the block is focused and so the sidebar stays open.

Seems like the tests have caught a potential bug here.

alexstine committed:
Cleanup code and try to fix more tests.

alexstine commented:
@talldan I have seen that console error before in other Blocks as well. Would be nice to know what is causing that.

I will try to track down what is happening with the focus function. Odd. I am not sure I will be able to fix it though as I never recalled it changing system focus which means I would have probably never noticed this anyway.

alexstine commented:
Hmm, have no idea why this would fail.

```
/**
 * Moves focus to the selected block.
 */
async function focusSelectedBlock() {
	// Ideally there shouuld be a UI way to do this. (Focus the selected block)
	await page.evaluate( () => {
		wp.data
			.dispatch( 'core/block-editor' )
			.selectBlock(
				wp.data
					.select( 'core/block-editor' )
					.getSelectedBlockClientId(),
				0
			);
	} );
}
```

It just checks to see in the store if the Block is selected or not. My code shouldn't be changing the interaction here unless it needs to focus the Block wrapper first. Weird.

alexstine commented:
```
Promise { <state>: "fulfilled", <value>: {…} }
​
<state>: "fulfilled"
​
<value>: Object { type: "SELECT_BLOCK", initialPosition: 0, clientId: "8d9a776e-8576-4e6c-ab09-724394007e83" }
​
<prototype>: Promise.prototype { … }
```

When I tried running the code in the console after Block insert, it works fine.

alexstine committed:
Fix snapshot.

alexstine committed:
Maybe fix another snapshot.

talldan commented:
> It just checks to see in the store if the Block is selected or not. My code shouldn't be changing the interaction here unless it needs to focus the Block wrapper first. Weird.

It does try to select the block. It isn't great that the end to end test does this programmatically, as these tests should be about replicating user actions. I think the reason behind it is that some blocks are hard to click on. It might be an option to change this to use List View.

> When I tried running the code in the console after Block insert, it works fine.

It can be tricky to reproduce manually. The test is inhumanly fast, it might be that it tries to click on the navigation block while it momentarily has no focusable elements or something like that. I know the navigation block needs to do some loading before it displays fully. Does your new code fallback to selecting the wrapper when a block has nothing in it?

alexstine commented:
@talldan If my code cannot find anything to focus, it should fallback. I guess I just don't understand what selectBlock() does. If I am checked out on latest trunk and I run the selectBlock() code from above, nothing happens for me. The Inserter remains open and my focus does not shift. Does selectBlock() only offer a visual change?

alexstine committed:
Make sure returning early doesn't prevent false positive of still having a chance to focus.

alexstine commented:
@talldan Ah, I was running selectBlock() from Browser Console. Of course focus would not change. 😞 I just pushed a commit which might make this more stable. If the Navigation Block was loading and does not have any elements to focus, the code would return early causing the other functions to not run. Later, I can work on how to deal with loading states and managing focus but that is a bit much for this already complicated PR with tests.

alexstine committed:
Update snapshots.

talldan commented:
> Ah, I was running selectBlock() from Browser Console

Yeah, I think running it from the browser console interferes with how it would normally work. I also noticed that the inserter sidebar won't close properly if focus is moved to the console to enter the command.

alexstine commented:
@talldan I am so close to getting the tests fixed. Can you spot the issues with the last 2 snapshots? I tried running the test-e2e -- -u command, but it seems it isn't going to fix these.

Thanks.

tellthemachines committed:
Fix test snapshot.

tellthemachines commented:
I pushed an update to the multi-block selection snapshot, so it should pass on CI now. The funny thing is it wasn't failing locally until I ran it in interactive mode. I couldn't manually reproduce the scenario where the end result is `4.` so no idea what might be happening there.

The preview test is a [known flaky](https://github.com/WordPress/gutenberg/issues/37976) so its failure is probably related to the changes in this branch. I couldn't reproduce it locally, in headless or interactive mode.

alexstine commented:
@tellthemachines how can this PR keep moving forward? Is there nothing that can be done until the flaky test is fixed? I guess if it isn't code related on my side since it seems to be failing elsewhere, it needs to be resolved in trunk?

tellthemachines commented:
@alexstine the flaky tests are a bit of a curse in this repo; they are often very hard to fix, so we end up getting around them by restarting the build until they pass.

I just did #39085 to try and fix the issue with the inserter in an empty block, and realised it might work as a fix for this issue too. The difference is that instead of setting focus to the first element in the placeholder, it just makes it possible to tab into the placeholder. This has the advantage of working on existing blocks as well as newly-inserted ones. Would welcome your feedback on it!

tellthemachines commented on packages/e2e-tests/specs/editor/various/block-deletion.test.js:
```diff
@@ -192,6 +192,7 @@ describe( 'deleting all blocks', () => {
 		// Add and remove a block.
 		await insertBlock( 'Image' );
 		await page.waitForSelector( 'figure[data-type="core/image"]' );
+		await page.keyboard.press( 'ArrowUp' );
```
Note that Arrow Up to select the block wrapper won't work if 'Contain text cursor inside block' is enabled, so it will be much harder to delete a block immediately after insertion.

alexstine replied:
@tellthemachines This didn't change though, right? If contains cursor option is selected, Up Arrow on latest trunk should not focus the wrapper either.

tellthemachines replied:
Sure, but with these changes the wrapper won't be focused after insertion anymore. This is only relevant when you add a block, immediately realise it's the wrong block, and want to quickly delete it. It would be nice to preserve that ability, but we should be able to find a workaround for it.

andrewserong replied:
> This is only relevant when you add a block, immediately realise it's the wrong block, and want to quickly delete it. It would be nice to preserve that ability, but we should be able to find a workaround for it.

I wonder if in a follow-up, we can look at in placeholder states that don't have an editable input field (e.g. the Columns placeholder state, but not the Twitter embed input field), and see if we can get backspace to still delete the block?

tellthemachines replied:
That won't work for Table though, because the first focusable field is an input 😅 
But we should be able to work something out. Definitely as a follow-up.

tellthemachines commented on packages/e2e-tests/specs/editor/various/toolbar-roving-tabindex.test.js:
```diff
@@ -20,18 +20,16 @@ async function expectLabelToHaveFocus( label ) {
 	await page.keyboard.press( 'ArrowRight' );
 	await expectLabelToHaveFocus( 'Move up' );
 	await page.keyboard.press( 'Tab' );
-	await expectLabelToHaveFocus( currentBlockLabel );
```
Is it a problem that we're now unable to focus the block wrapper on a new block? Since it has a "document" role I'd expect it to still be focusable.

The other thing to consider is we're introducing a change in how focus behaves only for newly-added blocks, so supposing we add a block, then move focus to another block or another part of the page, and then move back again, the behaviour will be different. This might create confusion, e.g. "why am I now able to focus the block wrapper when I wasn't before?"

alexstine replied:
@tellthemachines In my testing, the block wrapper can still gain focus if you use Up Arrow from an element. Not sure why it doesn't work in some tests.

ellatrix replied:
Why do we have to remove this check? Can we still test that moving focus from the toolbar to the content is still possible?

alexstine replied:
@ellatrix I changed this because landing on the first element inside the wrapper is no longer the aria-label of the block wrapper so the test failed. I tried to still ensure moving focus to toolbar and back worked but if you press Tab from toolbar, wrapper is skipped.

alexstine replied:
@ellatrix Rewrote the test to be less confusing. Hopefully better structure and comment helps. This test is lacking on comments as it is. Something to improve in the future.

alexstine replied:
Another update. I could never make the tests pass. Is it fine to try to improve the tests in a follow-up PR? Essentially I need to implement some type of function for content editable blocks vs. non-content editable blocks. Content editable blocks can have the aria-label parsed directly but non-content editable blocks cannot since Up Arrow is now required to focus the block wrapper. In my opinion, these tests are still good since it checks to see if the toolbar is still accessible via Shift+Tab and Arrow keys.

This PR is getting far too big for me to keep good track of as well so I'd like to split this off soon.

alexstine replied:
Oh, I think my E2E tests were fine but trunk is currently failing. 👎 I wish I would have checked that sooner. I wondered why they were passing locally... 😕

tellthemachines commented on packages/e2e-tests/specs/editor/various/toolbar-roving-tabindex.test.js:
```diff
@@ -64,53 +62,47 @@ describe( 'Toolbar roving tabindex', () => {
 		await insertBlock( 'Table' );
-		await testBlockToolbarKeyboardNavigation( 'Block: Table', 'Table' );
-		// Move focus to the first toolbar item
-		await page.keyboard.press( 'Home' );
```
Does pressing the 'Home' key no longer take us to the beginning of the toolbar? I'm not able to test this PR on Windows as gutenberg.run doesn't seem to be working.

alexstine replied:
@tellthemachines It did in my testing. That part should not be effected. The only things I changed around this were taking in to account not reading the aria-label from the block wrapper as focus was skipped.

tellthemachines committed:
Update tests with focus order changes.

tellthemachines reviewed (commented):
Thanks for your continued efforts on this PR! I left a few notes below. I think the main issue here is we're changing the tab sequence when the block is first added. Could we perhaps make it so the block wrapper can still be tabbed to?

I just pushed a few updates to the tests, including changes to undo a couple of snapshot diffs that shouldn't have changed. I'm not too sure about the Writing Flow 'should not have a dead zone above an aligned block' test, as its steps are very hard to understand and there are no comments to help guide us. I'm not even sure what the "dead zone" it's meant to be testing means 😅

alexstine committed:
Make the implementation more stable. Fix the List View focus bug.

alexstine commented on packages/block-editor/src/components/block-list/use-block-props/use-focus-first-element.js:
```diff
@@ -52,6 +52,16 @@ function useInitialPosition( clientId ) {
 	);
 }
 
+function isFormElement( element ) {
```
I borrowed this function from writing-flow. Should probably throw this somewhere else in the future.

alexstine commented on packages/block-editor/src/components/list-view/block.js:
```diff
@@ -106,6 +106,7 @@ function ListViewBlock( {
 	const selectEditorBlock = useCallback(
 		( event ) => {
 			selectBlock( event, clientId );
+			event.preventDefault();
```
Had to add this because if focus was placed on a button, the button would get clicked causing an unprovoked action. E.g.
1. Insert a Table block.
2. Select it from the List View.
3. The Create Table button gets clicked on block selection button click.
4. Adding this stops the event after selection.

alexstine reviewed (commented):
Highlighting some points.

alexstine committed:
Use target everywhere. Add a check for Block Inserter trigger.

alexstine commented on packages/block-editor/src/components/block-list/use-block-props/use-focus-first-element.js:
```diff
@@ -96,6 +106,25 @@ export function useFocusFirstElement( clientId ) {
+		// Check to see if Block contains focussable element before a generic caret insert.
+		if ( ! target.getAttribute( 'contenteditable' ) ) {
+			const focusElement = focus.tabbable.findNext( target );
+			// Ensure is not Add block/Block Inserter trigger
```
Not sure how else to check for this. Not clean, I know.

alexstine reviewed (commented):
One more note...

alexstine commented:
Right now, I think I've got this working really well. Implementation still probably needs some help but I am happy with the progress I made tonight. The List View bug was the nicest one to take care of. Hopefully the E2E tests will agree.

alexstine committed:
Merge branch 'trunk' of github.com:wordpress/gutenberg into try/accessible-block-content-dialog

alexstine committed:
Use Tab to exit the toolbar so blocks could be navigated. Only seems to be a partial fix for this E2E test failure.

tellthemachines committed:
Fix failing test and update snapshot.

ellatrix commented:
Hi @alexstine! I'm not sure I fully understand what this PR is doing. Could you add a new e2e test that illustrates why the change is needed? So an e2e test that would fail in trunk and pass for this branch.

alexstine commented:
@ellatrix I am not sure where to add the test at this point and I'm not sure I'm the right person for that job. Using Puppeteer is tricky for me without vision. I can try to mock the test, but not sure of positive results or not.

Here is what changed.

Before:
1. Open Gutenberg.
2. Insert some blocks such as Columns, Group, Navigation, Image, etc.
3. Open the List View.
4. Select the block.
5. Focus is placed on the block wrapper.

After:
1. Open Gutenberg.
2. Insert some blocks such as Columns, Group, Navigation, Image, etc.
3. Open the List View.
4. Select the block.
5. Instead of having focus placed on the block wrapper which doesn't make sense for most users, focus is placed on the first element that can receive focus. In the case of the image block, that would be the Upload button.
6. You can still navigate back to the block wrapper by using Up Arrow key.

Does this make sense?

Thanks.

alexstine committed:
Fix group block still places focus on Add block button. Add some E2E tests.

alexstine committed:
Merge branch 'trunk' of github.com:wordpress/gutenberg into try/accessible-block-content-dialog

alexstine committed:
Merge branch 'try/accessible-block-content-dialog' of github.com:wordpress/gutenberg into try/accessible-block-content-dialog

alexstine committed:
Fix code comment.

alexstine committed:
Fix bugs.

alexstine committed:
Merge branch 'trunk' of github.com:wordpress/gutenberg into try/accessible-block-content-dialog

alexstine committed:
Formatting fixes.

alexstine committed:
Update to use class name.

alexstine commented on packages/block-editor/src/components/block-list/use-block-props/use-focus-first-element.js:
```diff
@@ -96,6 +106,25 @@ export function useFocusFirstElement( clientId ) {
+		if ( ! target.getAttribute( 'contenteditable' ) ) {
+			const focusElement = focus.tabbable.findNext( target );
+			// Ensure is not block inserter trigger, don't want to focus that in the event of the group block which doesn't contain any other focussable elements.
+			const skipBlockInserterTrigger = target.classList.contains(
```
This has been something of a terrible pain. For some reason, hasAttribute/getAttribute throw errors on tests so I gave up and just used classList. I attempted to get it working with aria-label but for whatever reason, this is way easier in tests than it is in practice.

alexstine committed:
Try again to fix my own test.

alexstine commented:
@ellatrix I have implemented the E2E tests. Does this help you understand what is going on?

BTW, this is the first couple E2E tests I've added so still learning... It is an interesting mindset to take a step back and mock how your functionality should work.

andrewserong reviewed (commented):
Thanks for all the great work here @alexstine! I'm not all that familiar with the focusing on first element behaviour, but this is testing nicely for me, and it does feel like a good improvement when inserting a Columns, Table, or Image block, and having focus be on the first button within the placeholder state.

My testing:

* Before: when I insert an image block, the wrapper receives focus.
* After: when I insert an image block, the Upload button receives focus.

The additional and updated e2e tests are appreciated, as it's quite a detailed interaction to get right! There was a good point made about it being harder to immediately delete a block after insertion, but I agree, I think there'll be ways to come up with a fix for that in a follow-up.

It'd be good to also get a final seal of approval from @ellatrix and / or @tellthemachines, but I just wanted to chime in that from my perspective I think this PR is in good shape! 🎉

ellatrix commented:
I'm mostly AFK until Thursday, but would like to review this before merge. I'll try to do it today or tomorrow or otherwise Thursday.

alexstine committed:
Merge branch 'trunk' of github.com:wordpress/gutenberg into try/accessible-block-content-dialog

alexstine commented:
@andrewserong In theory, it should be possible. The only thing I never liked about this method of removing a block is focus gets super lost so this may give us a good opportunity to fix both issues.
1. Allow Backspace key to remove the block even if not on block wrapper for non-content editable blocks.
2. On block removal, make sure focus is dropped in a place that makes sense instead of leaving screen reader users in the void.

ellatrix reviewed (approved):
To me this seems fine. I was considering a similar behaviour in #33494 and the code will have to be received as part of that PR. I generally agree that focus should move into a placeholder when it is inserted.

alexstine committed:
Merge branch 'trunk' of github.com:wordpress/gutenberg into try/accessible-block-content-dialog

alexstine committed:
Try again to rewrite toolbar roving tabindex now that E2E is working.

alexstine committed:
Better code commenting.

alexstine commented on packages/e2e-tests/specs/editor/various/toolbar-roving-tabindex.test.js:
```diff
@@ -13,11 +13,17 @@ async function focusBlockToolbar() {
+	let ariaLabel = await page.evaluate( () =>
+		document.activeElement.getAttribute( 'aria-label' )
+	);
+	// If the labels don't match, try pressing Up Arrow to focus the block wrapper in non-content editable block.
```
I think this is way simpler compared to the modifications that were here and tests are still green. Seems logical to me. This whole file still is in desperate need of commenting.

tellthemachines reviewed (approved):
This is working really well 🎉 - tested by adding Image, Table and Columns blocks. Also tested moving focus from the List View into those blocks, and duplicating an Image block from the List View. In all cases, focus lands on the first focusable element within the block placeholder.

Updates to existing tests look good, and thanks for adding some new tests for this behaviour!
